lib/mozart_fetcher/fetcher.ex

```diff
@@ -1,30 +1,52 @@
 defmodule MozartFetcher.Fetcher do
-  alias MozartFetcher.{Component, TimeoutParser}
+  alias MozartFetcher.{Component, TimeoutParser, Envelope}
 
   use ExMetrics
 
+  @timeout_buffer 50
+  @max_concurrency 50
+
   def process([]) do
     ExMetrics.increment("error.empty_component_list")
     Stump.log(:error, %{message: "Error cannot process empty component list"})
     {:error}
   end
 
-  def process(components) do
+  def process(configs, buffer \\ @timeout_buffer) do
     ExMetrics.timeframe "function.timing.fetcher.process" do
-      max_timeout =
-        TimeoutParser.max(components) +
-          Application.get_env(:mozart_fetcher, :additional_task_await_timeout)
+      max_timeout = TimeoutParser.max(configs) + buffer
+
+      stream_opts = [timeout: max_timeout,
+                     on_timeout: :kill_task,
+                     max_concurrency: @max_concurrency]
 
-      components
+      configs
       |> Enum.with_index()
-      |> Enum.map(&Task.async(fn -> Component.fetch(&1) end))
-      |> Enum.map(&Task.await(&1, max_timeout))
-      |> decorate_response
+      |> Task.async_stream(&Component.fetch/1, stream_opts)
+      |> zip(configs)
+      |> decorate_response()
       |> Jason.encode!()
     end
   end
 
   defp decorate_response(envelopes) do
     %{components: envelopes}
   end
+
+  defp zip(responses, configs) do
+    for {response, config} <- Enum.zip(responses, configs) do
+      handle(response, config.id)
+    end
+  end
+
+  defp handle({:ok, result}, _id), do: result
+
+  defp handle({state, reason}, id) do
+    Stump.log(:error, %{message: "Component Process Error",
+                        state: state,
+                        reason: reason,
+                        id: id})
+
+    %{envelope: %Envelope{}, id: id, index: nil, status: nil}
```

Comment:
This would allow us to put a status, but we can think about this another time if needed.

Reply:
I could easily add a bit of logic which will inject a 408 id reason is `:timeout` otherwise 500.

Reply:
I'm doing that now